### PR TITLE
(TK-160) fix remaining "valid conf" parser tests

### PR DESCRIPTION
**Pull request by cprice404**

### CODE REVIEW

File: lib/hocon/impl/parseable.rb [hunk 1]
```diff
@@ -122,55 +166,43 @@ def parse_value(base_options = nil)
...
   end
 
-  def post_construct(base_options)
-    @initial_options = fixup_options(base_options)
-    @include_context = Hocon::Impl::SimpleIncludeContext.new(self)
-    if @initial_options.origin_description
-      @initial_origin = Hocon::Impl::SimpleConfigOrigin.new_simple(@initial_options.origin_description)
-    else
-      @initial_origin = create_origin
+  def parse_document(base_options = nil)
```
Comment (fpringvaldsen) [inline lib/hocon/impl/parseable.rb]: Should we port over the comments in this method as well?

File: lib/hocon/impl/parseable.rb [hunk 2]
```diff
@@ -179,11 +211,23 @@ def parse_value_from_origin(origin, final_options)
...
+
+    # after reader() we will have loaded the Content-Type
+    content_type = content_type()
+
+    options_with_content_type = nil
+    if !(content_type.nil?)
+      if Hocon::Impl::ConfigImpl.trace_loads_enabled && (! final_options.get_syntax.nil?)
+        self.class.trace("Overriding syntax #{final_options.get_syntax} with Content-Type which specified #{content-type}")
+      end
+      options_with_content_type = final_options.set_syntax(content-type)
```
Comment (fpringvaldsen) [inline lib/hocon/impl/parseable.rb]: Should be `content_type`

File: lib/hocon/impl/parseable.rb [hunk 3]
```diff
@@ -191,4 +235,293 @@ def raw_parse_value_from_io(io, origin, final_options)
...
+  def raw_parse_document(origin, final_options)
+    reader = reader(final_options)
+    content_type = content_type()
+
+    options_with_content_type = nil
+    if !(content_type.nil?)
+      if Hocon::Impl::ConfigImpl.trace_loads_enabled && (! final_options.get_syntax.nil?)
+        self.class.trace("Overriding syntax #{final_options.get_syntax} with Content-Type which specified #{content-type}")
+      end
+      options_with_content_type = final_options.set_syntax(content-type)
```
Comment (fpringvaldsen) [inline lib/hocon/impl/parseable.rb]: Same here

**Comment by fpringvaldsen** [comment]
I just had a couple of minor comments. I'm :+1: otherwise. Also, looks like this needs a rebase.

**Comment by cprice404** [comment]
rebased and added a commit that fixes the things @fpringvaldsen pointed out.

**Comment by fpringvaldsen** [comment]
:+1:

File: lib/hocon/config_error.rb [hunk 4]
```diff
@@ -32,13 +32,8 @@ class ConfigParseError < Hocon::ConfigError
...
   class ConfigWrongTypeError < Hocon::ConfigError
-    # If you need the 4 or 5-arg constructor version, use 'construct' instead.
-    def initialize(origin, message, cause = nil)
-      super(origin, message, cause)
-    end
-
-    def self.construct(origin, path, expected, actual, cause = nil)
-      self.class.new(origin, "#{path} has type #{actual} rather than #{expected}", cause)
+    def self.with_expected_actual(origin, path, expected, actual, cause = nil)
+      ConfigWrongTypeError.new(origin, "#{path} has type #{actual} rather than #{expected}", cause)
```
Comment (KevinCorcoran) [inline lib/hocon/config_error.rb]: I'm fine with this change but curious as to the reason ... I thought the existing code here (which I just wrote) would be sufficient?

File: lib/hocon/impl/origin_type.rb [hunk 5]
```diff
@@ -7,5 +7,9 @@ module Hocon::Impl::OriginType
...
   FILE = 1
   #URL = 2
-  #RESOURCE = 3
-end
\ No newline at end of file
+  # We don't actually support loading from the classpath / loadpath, which is
+  # what 'RESOURCE' is about in the upstream library.  However, some code paths
+  # still flow through our simplistic implementation of `ParseableResource`, so
+  # we need this constant.
+  RESOURCE = 3
```
Comment (KevinCorcoran) [inline lib/hocon/impl/origin_type.rb]: yikes

File: lib/hocon/config_error.rb (hunk 4 above)
Comment (cprice404) [inline lib/hocon/config_error.rb]: It was a merge conflict, I had already made the same change.

File: lib/hocon/impl/parseable.rb [hunk 6]
```diff
@@ -10,77 +11,104 @@
...
     end
+  end
 
+  # the general idea is that any work should be in here, not in the
+  # constructor, so that exceptions are thrown from the public parse()
+  # function and not from the creation of the Parseable.
+  # Essentially this is a lazy field. The parser should close the
+  # reader when it's done with it.
+  #{//}# ALSO, IMPORTANT: if the file or URL is not found, this must throw.
+  #{//}# to support the "allow missing" feature.
```
Comment (KevinCorcoran) [inline lib/hocon/impl/parseable.rb]: :circus_tent:

**Comment by KevinCorcoran** [comment]
It looks like `Parseable` is missing `parseConfigDocument` ... not sure if this intended to be a total port of that class or not.

**Comment by cprice404** [comment]
that configDocument stuff will get dealt with when we port @fpringvaldsen 's stuff down so i'm not too worried about it.

File: lib/hocon/impl/parseable.rb [hunk 7]
```diff
@@ -191,4 +241,293 @@ def raw_parse_value_from_io(io, origin, final_options)
...
+        sibling = Parseable.relative_to(@input, filename)
+      end
+      if sibling.nil?
+        nil
+      elsif sibling.exists?
+        self.class.trace("#{sibling" exists, so loading it as a file"}")
+        Parseable.new_file(sibling, options.set_origin_description(nil))
+      else
+        self.class.trace("#{sibling} does not exist, so trying it as a resource")
+        super(filename)
```
Comment (KevinCorcoran) [inline lib/hocon/impl/parseable.rb]: Is this correct?  The java version is `return super.relativeTo(filename);`

Hmm, is that what `super

Comment (cprice404) [inline lib/hocon/impl/parseable.rb]: yep

**Pull request merged**